### PR TITLE
add HMCForecaster

### PR DESCRIPTION
**Pull request by fehiepsi**
This PR supports HMC for Forecaster.

TODO:
- [x] add test

### CODE REVIEW

File: pyro/contrib/forecast/forecaster.py [hunk 1]
```diff
@@ -283,3 +284,64 @@ def forward(self, data, covariates, num_samples):
...
+        for name, node in list(self._trace.nodes.items()):
+            if name not in samples:
+                del self._trace.nodes[name]
+            else:
+                sample = samples[name]
+                node['value'] = sample.reshape(
+                    (self.num_samples,) + (1,) * (node['value'].dim() - sample.dim()) + sample.shape[1:])
+
+    @torch.no_grad()
+    def forward(self, data, covariates):
```
Comment (fehiepsi) [inline pyro/contrib/forecast/forecaster.py]: Do we need to support `num_samples` (by subsampling from mcmc samples) here? I recall that @martinjankowiak does not like that idea because MCMC samples are expensive.

Comment (fritzo) [reply pyro/contrib/forecast/forecaster.py]: I would like to keep the interfaces identical, so yes I'd like to keep a `num_samples` kwarg. This is also required for `crps` evaluation and plotting quantiles.

Can we not simply reuse the existing samples?

File: pyro/contrib/forecast/forecaster.py [hunk 2]
```diff
@@ -283,3 +284,64 @@ def forward(self, data, covariates, num_samples):
...
+        For models not using covariates, pass a shaped empty tensor
+        ``torch.empty(duration, 0)``.
+    :type covariates: ~torch.Tensor
+
+    :param int num_warmup: number of MCMC warmup steps.
+    :param int num_samples: number of MCMC samples.
+    :param int num_chains: number of parallel MCMC chains.
+    """
+    def __init__(self, model, data, covariates=None, *,
+                 num_warmup=1000, num_samples=1000, num_chains=1):
```
Comment (martinjankowiak) [inline pyro/contrib/forecast/forecaster.py]: can we please add support for more hmc options? `full_mass`, `max_tree_depth`, maybe jit?

Comment (fehiepsi) [reply pyro/contrib/forecast/forecaster.py]: Sure, I'll add them. `jit` is important btw, I'll enable it by default.

File: pyro/contrib/forecast/forecaster.py (hunk 1 above)
Comment (fehiepsi) [reply pyro/contrib/forecast/forecaster.py]: So I'll support that.

> Can we not simply reuse the existing samples?

In other frameworks, they use subsampling (Stan, Rlgt, PyMC3).

Comment (fehiepsi) [reply pyro/contrib/forecast/forecaster.py]: @fritzo I just support `num_samples` with subsampling.

File: tests/contrib/forecast/test_forecaster.py [hunk 3]
```diff
@@ -103,17 +103,24 @@ def model(self, zero_data, covariates):
 @pytest.mark.parametrize("t_obs", [1, 7])
 @pytest.mark.parametrize("t_forecast", [1, 3])
 @pytest.mark.parametrize("batch_shape", [(), (4,), (3, 2)], ids=str)
-@pytest.mark.parametrize("cov_dim", [0, 1, 6])
+@pytest.mark.parametrize("cov_dim", [0, 1, 5])
 @pytest.mark.parametrize("obs_dim", [1, 2])
 @pytest.mark.parametrize("dct_gradients", [False, True])
 @pytest.mark.parametrize("Model", [Model0, Model1, Model2, Model3, Model4])
-def test_smoke(Model, batch_shape, t_obs, t_forecast, obs_dim, cov_dim, dct_gradients):
+@pytest.mark.parametrize("engine", ["svi", "hmc"])
```
Comment (fehiepsi) [inline tests/contrib/forecast/test_forecaster.py]: `hmc` test is slow (~3m in my system) :(

Comment (fritzo) [reply tests/contrib/forecast/test_forecaster.py]: can we specify fewer samples and shorter warmup etc. to speed it up?

File: pyro/contrib/forecast/forecaster.py [hunk 4]
```diff
@@ -283,3 +284,79 @@ def forward(self, data, covariates, num_samples):
...
+    :param int num_warmup: number of MCMC warmup steps.
+    :param int num_samples: number of MCMC samples.
+    :param int num_chains: number of parallel MCMC chains.
+    :param bool dense_mass: a flag to decide if mass matrix is dense or diagonal.
+        Default to False.
+    :param bool jit_compile: whether to use the PyTorch JIT to trace the log
+        density computation, and use this optimized executable trace in the
+        integrator. Default to True.
+    :param int max_tree_depth: Max depth of the binary tree created during the doubling
+        scheme of NUTS sampler. Default to 10.
```
Comment (fritzo) [inline pyro/contrib/forecast/forecaster.py]: nit: could you add a sphinx link to the NUTS class?

File: tests/contrib/forecast/test_forecaster.py (hunk 3 above)
Comment (fehiepsi) [reply tests/contrib/forecast/test_forecaster.py]: awesome! Time is reduced to 30s now. Thanks a lot!

**Comment by fehiepsi** [comment]
@fritzo This does not add HMCForecaster to backtest. Could you suggest an api for it in `backtest`? I'll fill details then.

**Comment by fritzo** [comment]
@fehiepsi I think we add a `forcaster_fn` kwarg to `backtest()` and then remove the `guide_fn` kwarg (because `guide_fn` can the be simulated by wrapping `Forecaster` in a smarter `forecaster_fn`). Because the interfaces are the same, `backtest()` should still work. You should be able to ignore the `warm_start` logic because `HMCForecaster` does not accept a `warm_start` kwarg.
```diff
diff --git a/pyro/contrib/forecast/evaluate.py b/pyro/contrib/forecast/evaluate.py
index 1b551bfd..8a5fe4d2 100644
--- a/pyro/contrib/forecast/evaluate.py
+++ b/pyro/contrib/forecast/evaluate.py
@@ -66,7 +66,8 @@ DEFAULT_METRICS = {
 }
 
 
-def backtest(data, covariates, model_fn, guide_fn=None, *,
+def backtest(data, covariates, model_fn, *,
+             forecaster_fn=Forecaster,
              metrics=None,
              transform=None,
              train_window=None,
@@ -162,9 +163,7 @@ def backtest(data, covariates, model_fn, guide_fn=None, *,
         train_data = data[..., t0:t1, :]
         train_covariates = covariates[..., t0:t1, :]
         model = model_fn()
-        guide = None if guide_fn is None else guide_fn()
-        forecaster = Forecaster(model, train_data, train_covariates,
-                                guide=guide, **forecaster_options)
+        forecaster = forecaster_fn(model, train_data, train_covariates,
+                                   **forecaster_options)
 
         # Forecast forward to testing window.
         test_covariates = covariates[..., t0:t2, :]
```

**Comment by fritzo** [comment]
@fehiepsi can you also add a half sentence to the forecasting overview docs with a sphinx link to the `HMCForecaster` class?
https://github.com/pyro-ppl/pyro/blob/dev/docs/source/contrib.forecast.rst

**Comment by fehiepsi** [comment]
Thanks, @fritzo ! I have incorporated your suggestions.

**Comment by fehiepsi** [comment]
@fritzo @martinjankowiak I run HMCForecaster for tutorial i in [this gist](https://gist.github.com/fehiepsi/8bc0cd43289a60e79c176d73912a7fb2). For some reasons, with `jit_compile=True`, we got an error with Model2 (the traceback can be seen in the above gist). So I set `jit_compile=False` by default, to investigate later. It seems that HMC gives better results, but it is slow.

File: pyro/contrib/forecast/forecaster.py [hunk 5]
```diff
@@ -283,3 +284,80 @@ def forward(self, data, covariates, num_samples):
...
+    :type data: ~torch.Tensor
+    :param covariates: A tensor of covariates with time dimension -2.
+        For models not using covariates, pass a shaped empty tensor
+        ``torch.empty(duration, 0)``.
+    :type covariates: ~torch.Tensor
+
+    :param int num_warmup: number of MCMC warmup steps.
+    :param int num_samples: number of MCMC samples.
+    :param int num_chains: number of parallel MCMC chains.
+    :param bool dense_mass: a flag to decide if mass matrix is dense or diagonal.
```
Comment (martinjankowiak) [inline pyro/contrib/forecast/forecaster.py]: nit: A flag that controls whether the mass matrix is dense or diagonal. Defaults to False.

File: pyro/contrib/forecast/forecaster.py [hunk 6]
```diff
@@ -283,3 +284,80 @@ def forward(self, data, covariates, num_samples):
...
+        Default to False.
+    :param bool jit_compile: whether to use the PyTorch JIT to trace the log
+        density computation, and use this optimized executable trace in the
+        integrator. Default to False.
```
Comment (martinjankowiak) [inline pyro/contrib/forecast/forecaster.py]: nit: Defaults

File: pyro/contrib/forecast/forecaster.py [hunk 7]
```diff
@@ -283,3 +284,80 @@ def forward(self, data, covariates, num_samples):
...
+    :param int max_tree_depth: Max depth of the binary tree created during the doubling
+        scheme of :class:`~pyro.infer.mcmc.nuts.NUTS` sampler. Default to 10.
```
Comment (martinjankowiak) [inline pyro/contrib/forecast/forecaster.py]: nit: "scheme of the"
nit: Defaults

File: pyro/contrib/forecast/forecaster.py [hunk 8]
```diff
@@ -283,3 +284,80 @@ def forward(self, data, covariates, num_samples):
...
+        with poutine.trace() as tr:
+            with pyro.plate("particles", 1, dim=-self.max_plate_nesting - 1):
+                model(data, covariates)
+
+        self._trace = tr.trace
+        self._samples = mcmc.get_samples()
+        self._num_samples = num_samples * num_chains
+        for name, node in list(self._trace.nodes.items()):
+            if name not in self._samples:
+                del self._trace.nodes[name]
```
Comment (martinjankowiak) [inline pyro/contrib/forecast/forecaster.py]: what is this doing?

File: pyro/contrib/forecast/forecaster.py [hunk 9]
```diff
@@ -283,3 +284,80 @@ def forward(self, data, covariates, num_samples):
...
+
+    @torch.no_grad()
+    def forward(self, data, covariates, num_samples):
+        assert data.size(-2) < covariates.size(-2)
+        assert self.max_plate_nesting >= 1
+        dim = -1 - self.max_plate_nesting
+
+        weights = torch.ones(self._num_samples, device=data.device)
+        indices = torch.multinomial(weights, num_samples, replacement=True)
```
Comment (martinjankowiak) [inline pyro/contrib/forecast/forecaster.py]: perhaps replacement should be False if `num_samples <= self._num_samples`?

**Comment by martinjankowiak** [comment]
@fehiepsi i can confirm this seems to give reasonable results on a univariate model i've been playing around with. nice work!

File: pyro/contrib/forecast/forecaster.py (hunk 8 above)
Comment (fritzo) [reply pyro/contrib/forecast/forecaster.py]: as I understand this is garbage collecting param statements to save memory

Comment (fehiepsi) [reply pyro/contrib/forecast/forecaster.py]: In addition, this is to remove sites such as residuals, which should not be replayed in forward method.

**Comment by martinjankowiak** [comment]
@fritzo ready to merge as far as i'm concerned
